### PR TITLE
[ML] Add `TensorflowPredictor` interface

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Adds interface for `TensorflowPredictor`.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        model_definition: A callable that returns a TensorFlow Keras model
+            to use for predictions.
+        preprocessor: A preprocessor used to transform data batches prior
+            to prediction.
+        model_weights: List of weights to use for the model.
+    """
+
+    def __init__(
+        self,
+        model_definition: Type[tf.keras.Model],
```
Comment (amogkam) [inline python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:27]: Should this be Callable[[], tf.keras.Model] to match what’s in the doc string?

File: python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            model_definition: A callable that returns a TensorFlow Keras model
+                to use. Model weights will be loaded from the checkpoint.
+        """
+        raise NotImplementedError
+
+    def predict(
+        self,
+        data: DataBatchType,
+        column_names: Optional[List[str]] = None,
+        feature_columns: Optional[List[str]] = None,
```
Comment (amogkam) [inline python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py]: Can also be List[int]?

Comment (Yard1) [inline python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:54]:
```suggestion
        feature_columns: Optional[List[str], List[int]] = None,
```

File: python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                ``TorchTrainer`` run.
+            model_definition: A callable that returns a TensorFlow Keras model
+                to use. Model weights will be loaded from the checkpoint.
+        """
+        raise NotImplementedError
+
+    def predict(
+        self,
+        data: DataBatchType,
+        column_names: Optional[List[str]] = None,
```
Comment (amogkam) [inline python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py]: I originally had this as well, but no longer necessary I feel. If columns don’t have names you can still specify them with int indices.

File: python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                ``TorchTrainer`` run.
+            model_definition: A callable that returns a TensorFlow Keras model
+                to use. Model weights will be loaded from the checkpoint.
+        """
+        raise NotImplementedError
+
+    def predict(
+        self,
+        data: DataBatchType,
+        feature_columns: Optional[List[str], List[int]] = None,
```
Comment (Yard1) [inline python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:53]:
```suggestion
        feature_columns: Optional[Union[List[str], List[int]]] = None,
```

File: python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+# TensorFlow model objects cannot be pickled, therefore we use
+# a callable that returns the model, instead of a model object
+# itself.
+
+
+class TensorflowPredictor(Predictor):
+    """A predictor for TensorFlow models.
+
+    Args:
+        model_definition: A callable that returns a TensorFlow Keras model
```
Comment (amogkam) [inline python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:18]: What are the tradeoffs with accepting a callable vs. the model itself? Is this something we should do for torch as well?

File: python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py (hunk 1 above)
Comment (Yard1) [reply python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:27]: I think this is equivalent. Type signifies an uninitialized class. I may clarify that further in the docstring

Comment (amogkam) [reply python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:27]: I think we should err on the side of clarity...when I see this I think I have to pass in a class. But from the docstring, it seems like I have to pass in a function. Which one should I be doing?

Comment (Yard1) [reply python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:27]: Anything that will return an initialized model after being called with no arguments. I can make it an union of the two?

Comment (amogkam) [reply python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:27]: yes, I think that would clarify it!

**Review by ericl — APPROVED** [review]

File: python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    @classmethod
+    def from_checkpoint(
+        cls,
+        checkpoint: Checkpoint,
+        model_definition: Union[Callable[[], tf.keras.Model], Type[tf.keras.Model]],
+    ) -> "TensorflowPredictor":
+        """Instantiate the predictor from a Checkpoint.
+
+        The checkpoint is expected to be a result of ``TorchTrainer``.
```
Comment (simon-mo) [inline python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:41]:
```suggestion
        The checkpoint is expected to be a result of ``TensorflowTrainer``.
```

File: python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+        Args:
+            checkpoint: The checkpoint to load the model and
+                preprocessor from. It is expected to be from the result of a
+                ``TorchTrainer`` run.
```
Comment (simon-mo) [inline python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:46]:
```suggestion
                ``TensorflowTrainer`` run.
```

File: python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        model_definition: A callable that returns a TensorFlow Keras model
+            to use for predictions.
+        preprocessor: A preprocessor used to transform data batches prior
+            to prediction.
+        model_weights: List of weights to use for the model.
+    """
+
+    def __init__(
+        self,
+        model_definition: Union[Callable[[], tf.keras.Model], Type[tf.keras.Model]],
```
Comment (simon-mo) [inline python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:27]: Q: how does this relate to SavedModel format?

Comment (Yard1) [reply python/ray/ml/predictors/integrations/tensorflow/tensorflow_predictor.py:27]: No support for it yet, though we could add it in the future

**Review by amogkam — APPROVED** [review]

**Pull request merged**